File: scripts/check-bank-drift.py
#!/usr/bin/env python3
"""
Bank catalog drift checker.

Compares the official FinTS institute CSV (from Deutsche Kreditwirtschaft)
against banks.zig to detect:
- Missing banks (have FinTS URL in CSV but not in our catalog)
- URL changes (same BLZ, different URL)
- Removed banks (in our catalog but no longer in CSV with a FinTS URL)

Usage:
    python3 scripts/check-bank-drift.py path/to/fints_institute.csv
    python3 scripts/check-bank-drift.py path/to/fints_institute.csv --json-out drift-report.json
"""

from __future__ import annotations

import argparse
import csv
import json
import re
from pathlib import Path
from typing import Dict, List, NamedTuple, Optional


class CsvBank(NamedTuple):
    blz: str
    bic: str
    name: str
    url: str


class ZigBank(NamedTuple):
    blz: str
    bic: str
    name: str
    url: str


def parse_official_csv(path: Path) -> Dict[str, CsvBank]:
    """Parse the official FinTS institute CSV.

    The CSV has semicolon-separated fields. Expected columns (by header name):
    - Bankleitzahl (BLZ)
    - BIC
    - Bezeichnung (bank name)
    - URL (FinTS endpoint)

    The file may use different column names across versions, so we try
    multiple known header variants.
    """
    banks: Dict[str, CsvBank] = {}

    # Try multiple encodings (official file is typically latin-1 or cp1252)
    content = None
    for enc in ("utf-8-sig", "cp1252", "latin-1", "utf-8"):
        try:
            content = path.read_text(encoding=enc)
            break
        except (UnicodeDecodeError, UnicodeError):
            continue

    if content is None:
        raise ValueError(f"Cannot decode {path} with any supported encoding")

    # Detect delimiter (semicolon or comma)
    first_line = content.split("\n", 1)[0]
    delimiter = ";" if ";" in first_line else ","

    reader = csv.DictReader(content.splitlines(), delimiter=delimiter)
    if reader.fieldnames is None:
        raise ValueError(f"No header found in {path}")

    # Normalize field names for matching
    field_map: Dict[str, str] = {}
    for f in reader.fieldnames:
        fl = f.strip().lower()
        if "bankleitzahl" in fl or fl == "blz":
            field_map["blz"] = f
        elif fl == "bic" or "bic" in fl:
            field_map["bic"] = f
        elif "bezeichnung" in fl or "name" in fl or "institut" in fl:
            field_map["name"] = f
        elif "url" in fl or "pin/tan" in fl or "fints" in fl.replace("-", "").replace(" ", ""):
            field_map["url"] = f

    if "blz" not in field_map:
        raise ValueError(
            f"Cannot find BLZ column. Available columns: {reader.fieldnames}"
        )

    for row in reader:
        blz = row.get(field_map.get("blz", ""), "").strip()
        if not blz or not blz.isdigit():
            continue

        # Pad BLZ to 8 digits
        blz = blz.zfill(8)

        url = row.get(field_map.get("url", ""), "").strip()
        bic = row.get(field_map.get("bic", ""), "").strip()
        name = row.get(field_map.get("name", ""), "").strip()

        # Only include banks with a FinTS URL
        if not url or not url.startswith("http"):
            continue

        # Normalize URL: strip trailing slashes for consistent comparison
        url = url.rstrip("/")

        # Keep first occurrence per BLZ (some CSVs have duplicates)
        if blz not in banks:
            banks[blz] = CsvBank(blz=blz, bic=bic, name=name, url=url)

    return banks


def parse_banks_zig(path: Path) -> Dict[str, ZigBank]:
    """Parse banks.zig to extract all makeBank entries."""
    text = path.read_text(encoding="utf-8")
    pattern = re.compile(
        r'makeBank\("(?P<blz>\d{8})",\s*"(?P<bic>[^"]*)",\s*"(?P<name>[^"]*)",\s*"(?P<url>[^"]*)"\)'
    )
    banks: Dict[str, ZigBank] = {}
    for m in pattern.finditer(text):
        blz = m.group("blz")
        banks[blz] = ZigBank(
            blz=blz,
            bic=m.group("bic"),
            name=m.group("name"),
            url=m.group("url").rstrip("/"),
        )
    return banks


def normalize_url(url: str) -> str:
    """Normalize URL for comparison (lowercase host, strip trailing slash)."""
    return url.rstrip("/").lower()


def check_drift(
    csv_banks: Dict[str, CsvBank], zig_banks: Dict[str, ZigBank]
) -> tuple[list, list, list]:
    """Compare CSV against banks.zig. Returns (missing, url_changes, removed)."""

    missing: List[Dict] = []
    url_changes: List[Dict] = []
    removed: List[Dict] = []

    # Banks in CSV but not in banks.zig
    for blz, cb in sorted(csv_banks.items()):
        if blz not in zig_banks:
            missing.append({
                "blz": blz,
                "bic": cb.bic,
                "name": cb.name,
                "url": cb.url,
            })
        else:
            zb = zig_banks[blz]
            if normalize_url(cb.url) != normalize_url(zb.url):
                url_changes.append({
                    "blz": blz,
                    "name": cb.name,
                    "old_url": zb.url,
                    "new_url": cb.url,
                })

    # Banks in banks.zig but not in CSV (may have been removed from FinTS)
    for blz, zb in sorted(zig_banks.items()):
        if blz not in csv_banks:
            removed.append({
                "blz": blz,
                "name": zb.name,
                "url": zb.url,
            })

    return missing, url_changes, removed


def print_report(
    missing: list, url_changes: list, removed: list,
    csv_count: int, zig_count: int,
) -> None:
    print(f"Official CSV: {csv_count} banks with FinTS URLs")
    print(f"banks.zig:    {zig_count} banks")
    print()

    if not missing and not url_changes and not removed:
        print("No drift detected. banks.zig is up to date.")
        return

    if missing:
        print(f"MISSING ({len(missing)} banks in CSV but not in banks.zig):")
        print("-" * 80)
        for b in missing[:30]:  # Show first 30
            print(f"  {b['blz']}  {b['name'][:40]:<40}  {b['url']}")
        if len(missing) > 30:
            print(f"  ... and {len(missing) - 30} more")
        print()

    if url_changes:
        print(f"URL CHANGES ({len(url_changes)} banks with different URLs):")
        print("-" * 80)
        for b in url_changes:
            print(f"  {b['blz']}  {b['name'][:40]}")
            print(f"    old: {b['old_url']}")
            print(f"    new: {b['new_url']}")
        print()

    if removed:
        print(f"REMOVED ({len(removed)} banks in banks.zig but not in CSV):")
        print("-" * 80)
        for b in removed[:30]:
            print(f"  {b['blz']}  {b['name'][:40]:<40}  {b['url']}")
        if len(removed) > 30:
            print(f"  ... and {len(removed) - 30} more")
        print()


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Check bank catalog drift between official CSV and banks.zig"
    )
    parser.add_argument(
        "csv_path",
        type=Path,
        help="Path to official FinTS institute CSV file",
    )
    parser.add_argument(
        "--json-out",
        type=Path,
        default=None,
        help="Write JSON drift report to this path",
    )
    args = parser.parse_args()

    if not args.csv_path.exists():
        print(f"Error: CSV file not found: {args.csv_path}")
        return

    repo_root = Path(__file__).resolve().parents[1]
    banks_zig_path = repo_root / "libwimg" / "src" / "banks.zig"

    if not banks_zig_path.exists():
        print(f"Error: banks.zig not found at {banks_zig_path}")
        return

    csv_banks = parse_official_csv(args.csv_path)
    zig_banks = parse_banks_zig(banks_zig_path)

    print(f"Parsed {len(csv_banks)} banks from CSV, {len(zig_banks)} from banks.zig")
    print()

    missing, url_changes, removed = check_drift(csv_banks, zig_banks)
    print_report(missing, url_changes, removed, len(csv_banks), len(zig_banks))

    if args.json_out:
        report = {
            "summary": {
                "csv_count": len(csv_banks),
                "zig_count": len(zig_banks),
                "missing_count": len(missing),
                "url_change_count": len(url_changes),
                "removed_count": len(removed),
            },
            "missing": missing,
            "url_changes": url_changes,
            "removed": removed,
        }
        args.json_out.parent.mkdir(parents=True, exist_ok=True)
        args.json_out.write_text(
            json.dumps(report, indent=2, ensure_ascii=False), encoding="utf-8"
        )
        print(f"JSON report written: {args.json_out}")


if __name__ == "__main__":
    main()

File: scripts/check-bank-drift.ts
#!/usr/bin/env bun
/**
 * Bank catalog drift checker.
 *
 * Compares the official FinTS institute CSV (from Deutsche Kreditwirtschaft)
 * against banks.zig to detect:
 * - Missing banks (have FinTS URL in CSV but not in our catalog)
 * - URL changes (same BLZ, different URL)
 * - Removed banks (in our catalog but no longer in CSV with a FinTS URL)
 *
 * Usage:
 *     bun scripts/check-bank-drift.ts path/to/fints_institute.csv
 *     bun scripts/check-bank-drift.ts path/to/fints_institute.csv --json-out drift-report.json
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { resolve, dirname } from "path";

interface CsvBank {
	blz: string;
	bic: string;
	name: string;
	url: string;
}

interface ZigBank {
	blz: string;
	bic: string;
	name: string;
	url: string;
}

function parseOfficialCsv(path: string): Map<string, CsvBank> {
	const banks = new Map<string, CsvBank>();

	// Try multiple encodings — official file is typically latin-1 or cp1252
	let content: string;
	try {
		const buf = readFileSync(path);
		// Try UTF-8 first, fall back to latin-1
		content = new TextDecoder("utf-8", { fatal: true }).decode(buf);
	} catch {
		const buf = readFileSync(path);
		content = new TextDecoder("latin1").decode(buf);
	}

	// Strip BOM
	if (content.charCodeAt(0) === 0xfeff) content = content.slice(1);

	const lines = content.split(/\r?\n/).filter(Boolean);
	if (lines.length === 0) throw new Error(`Empty file: ${path}`);

	// Detect delimiter
	const delimiter = lines[0].includes(";") ? ";" : ",";
	const headers = lines[0].split(delimiter).map((h) => h.trim());

	// Map header names to column indices
	let blzIdx = -1;
	let bicIdx = -1;
	let nameIdx = -1;
	let urlIdx = -1;

	for (let i = 0; i < headers.length; i++) {
		const h = headers[i].toLowerCase();
		if (h.includes("bankleitzahl") || h === "blz") blzIdx = i;
		else if (h === "bic" || h.includes("bic")) bicIdx = i;
		else if (h.includes("bezeichnung") || h.includes("name") || h.includes("institut"))
			nameIdx = i;
		else if (h.includes("url") || h.includes("pin/tan") || h.replace(/[-\s]/g, "").includes("fints"))
			urlIdx = i;
	}

	if (blzIdx === -1) {
		throw new Error(`Cannot find BLZ column. Available: ${headers.join(", ")}`);
	}

	for (let i = 1; i < lines.length; i++) {
		const cols = lines[i].split(delimiter).map((c) => c.trim());
		let blz = cols[blzIdx] ?? "";
		if (!blz || !/^\d+$/.test(blz)) continue;
		blz = blz.padStart(8, "0");

		const url = (cols[urlIdx] ?? "").trim();
		if (!url || !url.startsWith("http")) continue;

		const normalizedUrl = url.replace(/\/+$/, "");
		const bic = (cols[bicIdx] ?? "").trim();
		const name = (cols[nameIdx] ?? "").trim();

		// Keep first occurrence per BLZ
		if (!banks.has(blz)) {
			banks.set(blz, { blz, bic, name, url: normalizedUrl });
		}
	}

	return banks;
}

function parseBanksZig(path: string): Map<string, ZigBank> {
	const text = readFileSync(path, "utf-8");
	const pattern = /makeBank\("(\d{8})",\s*"([^"]*)",\s*"([^"]*)",\s*"([^"]*)"\)/g;
	const banks = new Map<string, ZigBank>();
	let m: RegExpExecArray | null;
	while ((m = pattern.exec(text)) !== null) {
		banks.set(m[1], {
			blz: m[1],
			bic: m[2],
			name: m[3],
			url: m[4].replace(/\/+$/, ""),
		});
	}
	return banks;
}

function normalizeUrl(url: string): string {
	return url.replace(/\/+$/, "").toLowerCase();
}

function checkDrift(csvBanks: Map<string, CsvBank>, zigBanks: Map<string, ZigBank>) {
	const missing: Array<{ blz: string; bic: string; name: string; url: string }> = [];
	const urlChanges: Array<{ blz: string; name: string; old_url: string; new_url: string }> = [];
	const removed: Array<{ blz: string; name: string; url: string }> = [];

	for (const [blz, cb] of [...csvBanks.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
		const zb = zigBanks.get(blz);
		if (!zb) {
			missing.push({ blz, bic: cb.bic, name: cb.name, url: cb.url });
		} else if (normalizeUrl(cb.url) !== normalizeUrl(zb.url)) {
			urlChanges.push({ blz, name: cb.name, old_url: zb.url, new_url: cb.url });
		}
	}

	for (const [blz, zb] of [...zigBanks.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
		if (!csvBanks.has(blz)) {
			removed.push({ blz, name: zb.name, url: zb.url });
		}
	}

	return { missing, urlChanges, removed };
}

function printReport(
	missing: ReturnType<typeof checkDrift>["missing"],
	urlChanges: ReturnType<typeof checkDrift>["urlChanges"],
	removed: ReturnType<typeof checkDrift>["removed"],
	csvCount: number,
	zigCount: number,
) {
	console.log(`Official CSV: ${csvCount} banks with FinTS URLs`);
	console.log(`banks.zig:    ${zigCount} banks\n`);

	if (!missing.length && !urlChanges.length && !removed.length) {
		console.log("No drift detected. banks.zig is up to date.");
		return;
	}

	if (missing.length) {
		console.log(`MISSING (${missing.length} banks in CSV but not in banks.zig):`);
		console.log("-".repeat(80));
		for (const b of missing.slice(0, 30)) {
			console.log(`  ${b.blz}  ${b.name.slice(0, 40).padEnd(40)}  ${b.url}`);
		}
		if (missing.length > 30) console.log(`  ... and ${missing.length - 30} more`);
		console.log();
	}

	if (urlChanges.length) {
		console.log(`URL CHANGES (${urlChanges.length} banks with different URLs):`);
		console.log("-".repeat(80));
		for (const b of urlChanges) {
			console.log(`  ${b.blz}  ${b.name.slice(0, 40)}`);
			console.log(`    old: ${b.old_url}`);
			console.log(`    new: ${b.new_url}`);
		}
		console.log();
	}

	if (removed.length) {
		console.log(`REMOVED (${removed.length} banks in banks.zig but not in CSV):`);
		console.log("-".repeat(80));
		for (const b of removed.slice(0, 30)) {
			console.log(`  ${b.blz}  ${b.name.slice(0, 40).padEnd(40)}  ${b.url}`);
		}
		if (removed.length > 30) console.log(`  ... and ${removed.length - 30} more`);
		console.log();
	}
}

// --- Main ---

const args = process.argv.slice(2);
let csvPath: string | null = null;
let jsonOut: string | null = null;

for (let i = 0; i < args.length; i++) {
	if (args[i] === "--json-out" && args[i + 1]) {
		jsonOut = args[++i];
	} else if (!args[i].startsWith("-")) {
		csvPath = args[i];
	}
}

if (!csvPath) {
	console.error("Usage: bun scripts/check-bank-drift.ts path/to/fints_institute.csv [--json-out report.json]");
	process.exit(1);
}

const repoRoot = resolve(dirname(new URL(import.meta.url).pathname), "..");
const banksZigPath = resolve(repoRoot, "libwimg/src/banks.zig");

const csvBanks = parseOfficialCsv(csvPath);
const zigBanks = parseBanksZig(banksZigPath);

console.log(`Parsed ${csvBanks.size} banks from CSV, ${zigBanks.size} from banks.zig\n`);

const { missing, urlChanges, removed } = checkDrift(csvBanks, zigBanks);
printReport(missing, urlChanges, removed, csvBanks.size, zigBanks.size);

if (jsonOut) {
	const report = {
		summary: {
			csv_count: csvBanks.size,
			zig_count: zigBanks.size,
			missing_count: missing.length,
			url_change_count: urlChanges.length,
			removed_count: removed.length,
		},
		missing,
		url_changes: urlChanges,
		removed,
	};
	const outPath = resolve(repoRoot, jsonOut);
	mkdirSync(dirname(outPath), { recursive: true });
	writeFileSync(outPath, JSON.stringify(report, null, 2));
	console.log(`JSON report written: ${outPath}`);
}
